desc/objectives/_coils.py

```diff
@@ -0,0 +1,513 @@
+        Loss function to apply to the objective values once computed. This loss function
+        is called on the raw compute value, before any shifting, scaling, or
+        normalization. Note: Has no effect for this objective.
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute jacobian matrix, either forward mode or reverse mode AD.
+        "auto" selects forward or reverse mode based on the size of the input and output
+        of the objective. Has no effect on self.grad or self.hess which always use
+        reverse mode and forward over reverse mode respectively.
+    grid : Grid, list, optional
+        Collocation grid containing the nodes to evaluate at. If list, has to adhere to
+        Objective.dim_f
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    def __init__(
+        self,
+        coil,
+        data_keys,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name=None,
+    ):
+        self._grid = grid
+        self._data_keys = data_keys
+        self._normalize = normalize
+        super().__init__(
+            things=[coil],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        # local import to avoid circular import
+        from desc.coils import CoilSet, MixedCoilSet, _Coil
+
+        is_mixed_coils = isinstance(self.things[0], MixedCoilSet)
+        is_coil_set = isinstance(self.things[0], CoilSet)
+
+        coils = tree_flatten(
+            self.things[0],
+            is_leaf=lambda x: isinstance(x, _Coil) and not isinstance(x, CoilSet),
+        )[0]
+
+        # if using single coil, make coils and grid a list so they can be
+        # used with tree_map
+        coils = [coils[0]] if not is_coil_set else coils
+        if not isinstance(self._grid, list) and self._grid is not None:
+            self._grid = [self._grid]
+
+        if self._grid is None:
+            get_grid = lambda x: LinearGrid(
+                N=2 * x.N + 5, NFP=getattr(x, "NFP", 1), endpoint=False
+            )
+            self._grid = [get_grid(coil) for coil in coils]
+        elif np.any([grid.num_rho > 1 or grid.num_theta > 1 for grid in self._grid]):
+            raise ValueError("Only use toroidal resolution for coil grids.")
+        elif isinstance(self._grid, numbers.Integral):
```

Comment:
After line 110, self._grid is always either a list or None, so I this will never trigger. I think this check needs to get moved first.
Something like 
```python
# first handle int or None
if isinstance(self._grid, numbers.Integral):
     self._grid = LinearGrid(N=self._grid, endpoint=False)
if self._grid is None:
    get_grid = lambda x: LinearGrid(
                N=2 * x.N + 5, NFP=getattr(x, "NFP", 1), endpoint=False
            )
    self._grid = [get_grid(coil) for coil in coils]   
# then cast to list 
if not isinstance(self._grid, (tuple, list)):
    self._grid = [self._grid]
# then do checks
errorif(any(grid.num_rho>1 ...)